stl/inc/format

```diff
@@ -385,15 +386,106 @@ _NODISCARD constexpr const _CharT* _Parse_arg_id(
     throw format_error("Invalid format string.");
 }
 
+inline int _Code_units_in_next_character(const char* _First, const char* _Last, const _Cvtvec& _Cvt) {
+    // Returns a count of the number of code units (characters) that compose the first multibyte
+    // character in [_First, _Last), or -1 if [_First, _Last) doesn't contain an entire encoded
+    // character or *_First is not a valid lead byte.
+    _STL_INTERNAL_CHECK(_First < _Last);
+
+    switch (_Cvt._Mbcurmax) {
+    default:
+        _ASSERTE(!"Bad number of encoding units for this code page");
+        [[fallthrough]];
+    case 1:
+        return 1; // all characters have only one code unit
+
+    case 2:
+        {
+            wchar_t _Wide;
+            mbstate_t _St{};
+            const auto _Len   = static_cast<size_t>(_Last - _First);
+            const int _Result = _Mbrtowc(&_Wide, _First, _Len, &_St, &_Cvt);
+            if (_Result > 0) {
+                return _Result;
+            } else if (_Result < 0) { // invalid or incomplete encoded character
+                return -1;
+            } else { // next code unit is '\0'
+                return 1;
+            }
+        }
+
+    case 4: // Assume UTF-8 (as does _Mbrtowc)

```

Comment:
Are non-UTF-8 4-byte encodings (e.g. GB18030/code page 54936) possible here?

Reply:
This function (which is effectively a partially-inlined call to `_Mbrtowc`, which implements the equivalent for `std::codecvt`) and `_Mbrtowc` certainly don't think so. I'd like to do better, but I'm not prepared to overhaul how our locale machinery deals with character encodings in this PR.

Background: `_Mbrtowc` is a wrapper around a call to `MultiByteToWideChar` which relies on the anachronistic assumption that any encoded character can be converted to a single UCS-2 code unit. This is clearly wrong and has been for many years, but it's the mechanism we have right now for determining the extent of multibyte characters. I think we can do better, especially now that ICU is available so we may not need to write and maintain enormous quantities of per-supported-encoding code.